File: app/src/main/java/br/com/levimendesestudos/avenuecode/activities/MainActivity.java
package br.com.levimendesestudos.avenuecode.activities;

import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;
import java.util.List;
import br.com.levimendesestudos.avenuecode.R;
import br.com.levimendesestudos.avenuecode.adapters.AddressesListAdapter;
import br.com.levimendesestudos.avenuecode.models.Address;
import br.com.levimendesestudos.avenuecode.mvp.contracts.MainMVP;
import br.com.levimendesestudos.avenuecode.mvp.presenter.MainPresenter;
import br.com.levimendesestudos.avenuecode.utils.KeyBoardUtil;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnEditorAction;

public class MainActivity extends BaseActivity implements MainMVP.View {

    @BindView(R.id.tietSearch)
    TextInputEditText tietSearch;
    @BindView(R.id.tilSearch)
    TextInputLayout tilSearch;
    @BindView(R.id.ibSearch)
    ImageButton ibSearch;
    @BindView(R.id.rvAddresses)
    RecyclerView rvAddresses;
    @BindView(R.id.pbLoading)
    ProgressBar pbLoading;
    @BindView(R.id.tvNoResults)
    TextView tvNoResults;

    MainPresenter mPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        //etSearch.setText("SpringField");

        configureList();

        mPresenter = new MainPresenter(this);
    }

    @Override
    public void setErrorAddress(boolean value, String msg) {
        tilSearch.setErrorEnabled(value);
        tilSearch.setError(msg);
    }

    @Override
    public void cleanList() {
        rvAddresses.setAdapter(null);
    }

    @Override
    public void showPbLoading() {
        pbLoading.setVisibility(View.VISIBLE);
    }

    @Override
    public void hidePbLoading() {
        pbLoading.setVisibility(View.GONE);
    }

    @Override
    public void showNoResults() {
        tvNoResults.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideNoResults() {
        tvNoResults.setVisibility(View.GONE);
    }

    @Override
    public String address() {
        return tietSearch.getText().toString();
    }

    @OnEditorAction(R.id.tietSearch)
    public boolean etSearchAction() {
        mPresenter.search();

        return true;
    }

    @Override
    public void hideKeyboard() {
        KeyBoardUtil.hide(ibSearch);
    }

    @OnClick(R.id.ibSearch)
    public void ibSearchClick() {
        mPresenter.search();
    }

    private void configureList() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);

        rvAddresses.setLayoutManager(layoutManager);
        rvAddresses.setItemAnimator(new DefaultItemAnimator());
        rvAddresses.setHasFixedSize(true);
    }

    @Override
    public void loadList(List<Address> list) {
        AddressesListAdapter adapter = new AddressesListAdapter(this, list);
        rvAddresses.setAdapter(adapter);
    }
}

File: app/src/main/java/br/com/levimendesestudos/avenuecode/activities/MainActivity.kt
package br.com.levimendesestudos.avenuecode.activities

import android.os.Bundle
import android.support.design.widget.TextInputEditText
import android.support.design.widget.TextInputLayout
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.KeyEvent
import android.view.View
import android.widget.ImageButton
import android.widget.ProgressBar
import android.widget.TextView
import br.com.levimendesestudos.avenuecode.R
import br.com.levimendesestudos.avenuecode.adapters.AddressesListAdapter
import br.com.levimendesestudos.avenuecode.models.Address
import br.com.levimendesestudos.avenuecode.mvp.contracts.MainMVP
import br.com.levimendesestudos.avenuecode.mvp.presenter.MainPresenter
import br.com.levimendesestudos.avenuecode.utils.KeyBoardUtil
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.OnEditorAction

class MainActivity : BaseActivity(), MainMVP.View {

    @BindView(R.id.tietSearch)
    internal var tietSearch: TextInputEditText? = null
    @BindView(R.id.tilSearch)
    internal var tilSearch: TextInputLayout? = null
    @BindView(R.id.ibSearch)
    internal var ibSearch: ImageButton? = null
    @BindView(R.id.rvAddresses)
    internal var rvAddresses: RecyclerView? = null
    @BindView(R.id.pbLoading)
    internal var pbLoading: ProgressBar? = null
    @BindView(R.id.tvNoResults)
    internal var tvNoResults: TextView? = null

    lateinit var mPresenter: MainPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)

        tietSearch!!.setText("SpringField")

        configureList()

        mPresenter = MainPresenter(this)
    }

    override fun setErrorAddress(value: Boolean, msg: String) {
        tilSearch!!.isErrorEnabled = value
        tilSearch!!.error = msg
    }

    override fun cleanList() {
        rvAddresses!!.adapter = null
    }

    override fun showPbLoading() {
        pbLoading!!.visibility = View.VISIBLE
    }

    override fun hidePbLoading() {
        pbLoading!!.visibility = View.GONE
    }

    override fun showNoResults() {
        tvNoResults!!.visibility = View.VISIBLE
    }

    override fun hideNoResults() {
        tvNoResults!!.visibility = View.GONE
    }

    override fun address(): String {
        return tietSearch!!.text.toString()
    }

    @OnEditorAction(R.id.tietSearch)
    fun etSearchAction(textView: TextView, i: Int, keyEvent: KeyEvent): Boolean {
        mPresenter.search()

        return true
    }

    override fun hideKeyboard() {
        KeyBoardUtil.hide(ibSearch!!)
    }

    @OnClick(R.id.ibSearch)
    fun ibSearchClick() {
        mPresenter.search()
    }

    private fun configureList() {
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = LinearLayoutManager.VERTICAL

        rvAddresses!!.layoutManager = layoutManager
        rvAddresses!!.itemAnimator = DefaultItemAnimator()
        rvAddresses!!.setHasFixedSize(true)
    }

    override fun loadList(list: List<Address>) {
        val adapter = AddressesListAdapter(this, list)
        rvAddresses!!.adapter = adapter
    }
}